x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregatorTests.java

```diff
@@ -0,0 +1,305 @@
+                assertThat(
+                    categorizationAggregation.getBuckets().get(1).getKeyAsString(),
+                    equalTo("Failed to shutdown error org.aaaa.bbbb.Cccc line caused by foo exception")
+                );
+                assertThat(((Max) categorizationAggregation.getBuckets().get(1).aggregations.get("max")).getValue(), equalTo(4.0));
+                assertThat(((Min) categorizationAggregation.getBuckets().get(1).aggregations.get("min")).getValue(), equalTo(4.0));
+                assertThat(((Avg) categorizationAggregation.getBuckets().get(1).aggregations.get("avg")).getValue(), equalTo(4.0));
+            }
+        }
+    }
+
+    private static void writeTestDocs(RandomIndexWriter w) throws IOException {
+        w.addDocument(
+            Arrays.asList(
+                new StoredField("_source", new BytesRef("{\"text\":\"Node 1 started\"}")),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 0)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField("_source", new BytesRef("{\"text\":\"Node 1 started\"}")),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 1)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField(
+                    "_source",
+                    new BytesRef("{\"text\":\"Failed to shutdown [error org.aaaa.bbbb.Cccc line 54 caused by foo exception]\"}")
+                ),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 0)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField(
+                    "_source",
+                    new BytesRef("{\"text\":\"Failed to shutdown [error org.aaaa.bbbb.Cccc line 54 caused by foo exception]\"}")
+                ),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 4)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField("_source", new BytesRef("{\"text\":\"Node 2 started\"}")),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 2)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField("_source", new BytesRef("{\"text\":\"Node 2 started\"}")),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 3)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField("_source", new BytesRef("{\"text\":\"Node 3 started\"}")),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 4)
+            )
+        );
+        w.addDocument(
+            Arrays.asList(
+                new StoredField("_source", new BytesRef("{\"text\":\"Node 3 started\"}")),
+                new SortedNumericDocValuesField(NUMERIC_FIELD_NAME, 5)
+            )
+        );
+    }
```

Comment:
I think it'd be nice to have a test with a couple thousand documents, preferably random. I'd be looking to see what happens when you end up with many segments, especially when `AggregatorTestCase#searchAndReduce` get `splitLeavesIntoSeparateAggregators = true` and you get lots of merging. It can find all kinds of funky corner cases.

It'd also be nice to see when `splitLeavesIntoSeparateAggregators = false` and you end up having to merge text categories. That *should* have caught that sub-aggs don't merge properly.

Reply:
I now have a test that runs over 5k docs. I also noticed that `testCase` randomizes `splitLeavesIntoSeparateAggregators` automatically. So I think I have this covered.